spec.md

```diff
@@ -0,0 +1,276 @@
+
+#### Response
+A 301 Temporary Redirect to `ipns://<key>[/<path>]`. The header `X-IPNS-Path` is set to the `/ipns/<key>[/<path>]`.
+
+### POST `/ipns/localhost[/<path>]`
+
+Behaves the same as `/ipns/<key>[/<path>]`.
+
+#### Query params
+- `key=some_name`: **Required**. This specifies the key you want by a name instead of by public key. If a key with this name doesn't exist, it will be created.
+
+#### Response
+The same as `/ipns/<key>[/<path>]`. Note this allows for extraction of the actual public key from future use and link sharing.
+
+### PUT `/ipns/<key>[/<path>]`
+
+The same as POST `/ipns/<key>[/<path>]`, but where the response body is the file to be uploaded, not an IPFS path. Multiple files can be uploaded using form data, in which case files will go inside the directory in the path.
+
+### PUT `/ipns/localhost[/<path>]`
+
+The same as POST `/ipns/localhost[/<path>]`, but where the response body is the file to be uploaded, not an IPFS path. Multiple files can be uploaded using form data, in which case files will go inside the directory in the path.
+
+### DELETE `/ipns/<key>[/<path>]`
+
+Remove content behind an IPNS name. The new content is pinned, and the previous content is unpinned.
+
+#### Response
+A 301 Temporary Redirect to `ipns://<key>[/<path>]`. The path in the redirect is the directory containing the deletion path. So if the path in the URL was `/ipns/<key>/path/to/file.ext`, the redirect will be to `ipns://<key>/path/to`.
+
+The header `X-IPNS-Path` is set to `/ipns/<key>[/<path>]`, with the same path logic as above.
+
+### DELETE `/ipns/localhost[/<path>]`
+
+Behaves the same as `/ipns/<key>[/<path>]`.
+
+#### Query params
+- `key=some_name`: **Required**. This specifies the key you want by a name instead of by public key. If a key with this name doesn't exist, it will be created.
+
+#### Response
+The same as `/ipns/<key>[/<path>]`. Note this allows for extraction of the actual public key from future use and link sharing.
+
+## Pubsub
+
+An experimental pubsub API is supported to give users direct access to sending and receiving pubsub messages.
+
+### GET `/pubsub/<topic>`
+
+#### Query params
+- `format`: can be `json`, `utf8`, `base64`, or not specified, which defaults to `base64`. This controls the decoding of pubsub message bytes
+
+#### Response
+
+The response is an [event stream](https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events#event_stream_format), so pubsub messages for that topic will be sent to the client as long as the connection remains open.
+
+Pubsub messages are encoded as events like this:
+
+```
+id: FuhB2orEIwE=
+data: {"from":"QmfULfLaRtF7Kg3ShjFdiYL97P7eEcyL4PcvQLhkzCPCeG","data":"aGVsbG8=","topics":["test"]}
+```
+
+The "message identifier" of the pubsub message is base64-encoded to form the `id:`. `data:` is always encoded as JSON, with the other three things known about a pubsub message.
```

Comment:
Can we use the prettified IDs method from `from` for the message ID as well?

Reply:
There are no methods for prettifying this data, and no standard or documentation that I can find on doing that.

The pubsub spec specifies that the message ID is a 64-bit big-endian uint. So decoded would look like `1652103550844330915` instead of what I put in the spec. If that's preferable, I can switch to that.

> The seqno field (optional) is a 64-bit big-endian uint that is a linearly increasing number that is unique among messages originating from each given peer. No two messages on a pubsub topic from the same peer should have the same seqno value, however messages from different peers may have the same sequence number. In other words, this number is not globally unique.

From https://github.com/libp2p/specs/tree/master/pubsub#the-message

Reply:
If there's no standard, whatever you have now should be fine too 👍